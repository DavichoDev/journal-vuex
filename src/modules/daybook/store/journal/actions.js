
export const loadEntries = async (/*{ commit }*/) => {

}

export const updateEntries = async (/*{ commit }*/) => {

}

export const createEntry = (/*{ commit }*/) => {
    
}