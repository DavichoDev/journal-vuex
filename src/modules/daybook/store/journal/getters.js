export const getEntriesByTerm = ( { entries } ) => ( term = '' ) => {
    
    if ( term.length === 0 ) { return entries } 
    return entries.filter( entry => entry.text.toLowerCase().includes( term.toLowerCase() ))
    
}

export const getEntryById = ({ entries }) => ( id = '' ) => {
    const entry = entries.find( entry => entry.id === id )
    if(!entry) return
    return {...entry}
}