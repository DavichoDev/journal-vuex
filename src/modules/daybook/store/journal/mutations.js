
export const setEntries = (state, entries) => {
    state.entries = [...state.entries, ...entries]
    state.isLoading = false;
}

export const updateEntry = ( state, entry ) => {
    const updatedEntries = [...state.entries]
        .map(ent => ent.id === entry.id ? entry : ent)
    state.entries = [...updatedEntries]
    state.isLoading = false;
}

export const addEntry = ( state, entry ) => {
    state.entries = [entry, ...state.entries]
}

export const deleteEntry = ( state, idEntry ) => {
    const newEntries = [...state.entries].filter(entry => entry.id !== idEntry )
    state.entries = [...newEntries]
}