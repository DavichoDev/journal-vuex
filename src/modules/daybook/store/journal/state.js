
export default () => ({
    isLoading: true,
    entries: [
        {
            id: new Date().getTime(),
            date: new Date().toDateString(),
            text: 'Lorem ipsum dolor sit amet consectetur adipisicing elit. Error, dolorem veritatis explicabo, voluptatem, est corrupti possimus eligendi illum suscipit.',
            picture: null
        },
        {
            id: new Date().getTime() + 1000,
            date: new Date().toDateString(),
            text: 'Error, dolorem veritatis explicabo, voluptatem, est corrupti possimus eligendi illum suscipit. Lorem ipsum dolor sit amet consectetur adipisicing elit.',
            picture: null
        },
        {
            id: new Date().getTime() + 2000,
            date: new Date().toDateString(),
            text: 'Est corrupti possimus eligendi illum suscipit. Ipsum dolor sit amet consectetur adipisicing elit lorem. Error, dolorem veritatis explicabo, voluptatem.',
            picture: null
        }
    ]
})
