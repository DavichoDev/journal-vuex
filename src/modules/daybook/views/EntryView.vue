<template>
    <template v-if="entry">
        <div class="entry-title p-2 d-flex justify-content-between">
            <div>
                <span class="text-success fs-3 fw-bold">{{ day }}</span>
                <span class="mx-1 fs-3">{{ month }}</span>
                <span class="mx-2 fs-4 fw-light">{{ yearDate }}</span>
            </div>
            <div>
                
                <input 
                    type="file"
                    @change="onSelectedImage"
                    ref="imageSelector"
                    v-show="false"
                    accept="image/png, image/jpeg"
                >

                <button
                    @click="onDeleteEntry" 
                    class="btn btn-danger mx-2">
                    Borrar
                    <i class="fa fa-trash-alt"></i>
                </button>
                <button 
                    @click="onSelectImage"
                    class="btn btn-primary mx-2">
                    Subir foto
                    <i class="fa fa-upload "></i>
                </button>
            </div>
        </div>
        
        <hr>
        <div class="d-flex flex-column px-3 h-75">
            <textarea
                v-model="entry.text"
                placeholder="¿Qué sucedio hoy?"
            ></textarea>
        </div>
    </template>

    <Fab 
         @on:click="onSaveEntry"
         :icon="'fa-save'"
    />

    <img
        v-if="entry.picture && !localImage" 
        :src="entry.picture" 
        alt="entry-picture"
        class="img-thumbnail"
    />

    <img
        v-if="localImage" 
        :src="localImage" 
        alt="entry-picture"
        class="img-thumbnail"
    />

</template>

<script>
import { defineAsyncComponent } from 'vue'
import { mapActions, mapGetters } from 'vuex'
import Swal from 'sweetalert2'
import uploadImage from '../helpers/uploadImage';

import getDayMonthYear from '../helpers/getDayMonthYear';

export default {
    name: 'EntryView',
    props:{
        id: {
            type: String,
            required: true
        }
    },
    components: {
        Fab: defineAsyncComponent(() => import('../components/Fab.vue'))
    },
    data(){
        return {
            entry: {},
            localImage: null,
            imgFile: null
        }
    },
    computed: {
        ...mapGetters('journal',['getEntryById']),
        day(){
            const { day } = getDayMonthYear( this.entry.date )
            return day
        },
        month(){
            const { month } = getDayMonthYear( this.entry.date )
            return month
        },
        yearDate(){
            const { yearDate } = getDayMonthYear( this.entry.date )
            return yearDate
        }
    },
    methods: {
        ...mapActions('journal', ['updateEntries', 'createEntry', 'deleteEntry']),
        loadEntry () {
            let entry;
            if (this.id === 'new') {
                entry = {
                    text: '',
                    date: new Date().getTime()
                }
            } else {
                entry = this.getEntryById(this.id)
                if ( !entry ) return this.$router.push({name: 'no-entry'})
            }
            this.entry = entry
        },
        async onSaveEntry(){

            new Swal({
                title: 'Espere por favor',
                allowOutsideClick: false,
            })
            Swal.showLoading()

            const picture = await uploadImage(this.imgFile)

            this.entry.picture = picture

            if ( this.entry.id ) {
                // Actualizar
                await this.updateEntries(this.entry)
            } else {
                // Crear una nueva entrada
                const id = await this.createEntry(this.entry)
                this.$router.push({name: 'entry', params: {id}})
            }
            this.imgFile = null
            Swal.fire('Guardado', 'Entrada registrada con éxito', 'success')

        },
        async onDeleteEntry(){
            const { isConfirmed } = await Swal.fire({
                title: '¿Estás seguro?',
                text: 'Una vez borrado, no se podrá recuperar.',
                showDenyButton: true,
                confirmButtonText: 'Si, estoy seguro'
            })

            if (isConfirmed) {
                Swal.fire({
                    title: 'Espere por favor',
                    allowOutsideClick: false
                })
                Swal.showLoading()
                await this.deleteEntry(this.entry.id)
                this.$router.push({name: 'no-entry'})

                Swal.fire('Eliminado', '', 'success')
            }
        },
        onSelectedImage({target}){
            const file = target.files[0]
            if (!file) {
                this.localImage = null
                this.imgFile = null
                return
            }
            // imgFile para enviar
            this.imgFile = file
            const fr = new FileReader()
            fr.onload = () => this.localImage = fr.result
            fr.readAsDataURL( file )
        },
        onSelectImage(){
            this.$refs.imageSelector.click()
        }
    },
    created(){
        this.loadEntry()
    },
    watch: {
        id(){
            this.loadEntry()
        }
    }
}
</script>

<style lang="scss" scoped>
textarea {
    font-size: 20px;
    border: none;
    height: 100%;
    resize: none;

    &:focus{
        outline: none;
    }
}
img {
    width: 200px;
    position: fixed;
    bottom: 150px;
    right: 20px;
    box-shadow: 0px 5px 10px rgba($color: #000000, $alpha: 0.2);
}
.entry-title{

}
</style>