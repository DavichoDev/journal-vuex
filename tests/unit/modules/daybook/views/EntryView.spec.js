import { createStore } from "vuex";
import { shallowMount } from "@vue/test-utils";
import EntryView from '@/modules/daybook/views/EntryView.vue';
import journal from '@/modules/daybook/store/journal';
import { journalState } from "../../../mock-data/test-state";
import Swal from "sweetalert2";

const createVuexStore = ( initialState ) => 
createStore({
    modules: {
        journal: {
            ...journal,
            state: {...initialState}
        }
    }
})

jest.mock('sweetalert2', () => ({
    fire: jest.fn(),
    showLoading: jest.fn(),
    close: jest.fn(),
}))

describe('EntryView.vue Tests', () => {

    const store = createVuexStore( journalState )
    store.dispatch = jest.fn()
    const $router = { push: jest.fn() }

    let wrapper
    beforeEach(() => {
        jest.clearAllMocks()
        wrapper = shallowMount( EntryView, {
            props: {
                id: '-MnwodvOg-vnNI7zQ3aK'
            },
            global: {
                mocks: {
                    $router 
                },
                plugins: [ store ]
            }
        })
    })


    test('Debe de sacar al usuario porque el ID no existe', () => {
        shallowMount( EntryView, {
            props: {
                id: 'Este id no existe en el Store'
            },
            global: {
                mocks: { $router },
                plugins: [ store ]
            }
        })
        expect( $router.push ).toHaveBeenCalledWith({name: 'no-entry'})
    })

    test('Debe de mostrar la entrada correctamente', () => {
        expect(wrapper.html()).toMatchSnapshot()
        expect( $router.push ).not.toHaveBeenCalled()
    })

    test('Debe de borrar la entrada y salir', (done) => {
        
        Swal.fire.mockReturnValueOnce(Promise.resolve({isConfirmed: true}))
        wrapper.find('.btn-danger').trigger('click')
        expect( Swal.fire ).toHaveBeenCalledWith({
            title: '¿Estás seguro?',
            text: 'Una vez borrado, no se podrá recuperar.',
            showDenyButton: true,
            confirmButtonText: 'Si, estoy seguro'
        })
        setTimeout(() => {
            expect( store.dispatch ).toHaveBeenCalledWith('journal/deleteEntry', '-MnwodvOg-vnNI7zQ3aK')
            expect( $router.push ).toHaveBeenCalled()
            done()
        }, 1);
    })
})