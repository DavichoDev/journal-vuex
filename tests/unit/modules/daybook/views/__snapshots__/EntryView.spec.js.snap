// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntryView.vue Tests Debe de mostrar la entrada correctamente 1`] = `
<div class="entry-title p-2 d-flex justify-content-between">
  <div><span class="text-success fs-3 fw-bold">7</span><span class="mx-1 fs-3">Noviembre</span><span class="mx-2 fs-4 fw-light">2021, Domingo</span></div>
  <div><input type="file" accept="image/png, image/jpeg" style="display: none;"><button class="btn btn-danger mx-2"> Borrar <i class="fa fa-trash-alt"></i></button><button class="btn btn-primary mx-2"> Subir foto <i class="fa fa-upload"></i></button></div>
</div>
<hr>
<div class="d-flex flex-column px-3 h-75"><textarea placeholder="¿Qué sucedio hoy?"></textarea></div>
<fab-stub icon="fa-save"></fab-stub><img src="https://res.cloudinary.com/davicho/image/upload/v1636332235/h5cs9xvw8xzjniwdiqbz.jpg" alt="entry-picture" class="img-thumbnail">
<!--v-if-->
`;
